evo_researcher/benchmark/agents.py

```diff
@@ -11,79 +12,109 @@
-    def __init__(self, model: str):
-        super().__init__(agent_name="olas")
+    def __init__(self, model: str, temperature: float, agent_name: str = "olas", max_workers: t.Optional[int] = None):
+        super().__init__(agent_name=agent_name, max_workers=max_workers)
         self.model = model
-
-    def research_and_predict(self, market_question: str) -> Prediction:
-        report = research_autonolas(
-            prompt=market_question,
-            engine=self.model,
-        )
-        return _make_prediction(
-            market_question=market_question, additional_information=report
-        )
-
+        self.temperature = temperature
+    def evaluate(self, market_question: str) -> EvalautedQuestion:
+        return evaluate_question(question=market_question)
+
+    def research(self, market_question: str) -> t.Optional[str]:
+        try:
+            return research_autonolas(
+                prompt=market_question,
+                engine=self.model,
+            )
+        except ValueError as e:
+            print(f"Error in OlasAgent's research: {e}")
+            return None
+        
+    def predict(self, market_question: str, researched: str) -> t.Optional[Prediction]:
+        try:
+            return _make_prediction(
+                market_question=market_question,
+                additional_information=researched,
+                engine=self.model,
+                temperature=self.temperature,
+            )
+        except ValueError as e:
+            print(f"Error in OlasAgent's predict: {e}")
+            return None
 
 class EvoAgent(AbstractBenchmarkedAgent):
-    def __init__(self, model: str):
-        super().__init__(agent_name="evo", max_workers=4)
+    def __init__(self, model: str, temperature: float, agent_name: str = "evo", max_workers: t.Optional[int] = None):
+        super().__init__(agent_name=agent_name, max_workers=max_workers)
         self.model = model
+        self.temperature = temperature
+
+    def evaluate(self, market_question: str) -> EvalautedQuestion:
+        return evaluate_question(question=market_question)
 
-    def research_and_predict(self, market_question: str) -> Prediction:
+    def research(self, market_question: str) -> t.Optional[str]:
         dotenv.load_dotenv()
         open_ai_key = os.getenv("OPENAI_API_KEY")
         tavily_key = os.getenv("TAVILY_API_KEY")
-        report, _ = research_evo(
-            goal=market_question,
-            openai_key=open_ai_key,
-            tavily_key=tavily_key,
-            model=self.model,
-        )
-        return _make_prediction(
-            market_question=market_question, additional_information=report
-        )
+        try:
+            report, _ = research_evo(
+                goal=market_question,
+                openai_key=open_ai_key,
+                tavily_key=tavily_key,
+                model=self.model,
+            )
+            return report
+        except ValueError as e:
```

Comment:
Wrapped in try/except so that it doesn't crash the benchmark if there is some unexpected completion from LLM.

Reply:
None will be saved in the cache and it won't rerun, if the agent didn't answer on the first time, do you think it would on the second? If so, we could add retry logic here.

Reply:
Would also be great to see a test for this

Reply:
> I guess you made this change because of flakiness from 3rd party apis?

Not always, sometimes it's due to malformed LLM completion. 3rd party apis like Tavily have retry decorator already, so if they fail after N retries, there is no point in rerunning the whole agent because of that.

> Would also be great to see a test for this

Added!

Reply:
Sorry, I meant a test for the case where:
- a benchmarker loads a cache for which there's a `None` agent prediction
- when the benchmarker runs, it re-runs that prediction and is observed getting a non-None result.

Reply:
> when the benchmarker runs, it re-runs that prediction

but currently, it doesn't rerun the prediction; if it failed the first time, why wouldn't it fail the second time? 🤔

Reply:
Well I guess we have 2 cases of 'None' predictions now:
1. The agent evaluates the question and decides not to answer
2. The agent tries to answer, but throws an exception (flaky api, malformed json)

For (1) it probably shouldn't try to re-run. For (2) I think it should.

Reply:
Yeah, I'll differentiate it as you proposed in the other comment; that's good to have, for sure.